modules/dhcp/providers/server/isc.rb

```diff
@@ -78,13 +78,13 @@ def parse_config_and_leases_for_records
             opts.merge!(parse_record_options(data))
           end
         end
-        begin
-          subnet = @service.find_subnet(opts[:ip])
-          next unless subnet
-          ret_val << Proxy::DHCP::Reservation.new(opts.merge(:subnet => subnet))
-        rescue Exception => e
-          logger.warn "skipped #{hostname} - #{e}"
+        if opts[:deleted]
+          ret_val.delete_if { |rec| rec.options[:hostname] == hostname }
+          next
         end
```

Comment:
This bit executes in polynomial time (and O(n*n) for the whole lease file), which isn't good, as lease file parsing is pretty slow already. You could replace the array here with a hashmap indexed by hostname, or introduce support for subnet-less reservations (pls. see my earlier comment).